File: android/src/main/java/com/gbwisx/msal_mobile/Results.java
package com.gbwisx.msal_mobile;

import com.google.gson.Gson;
import com.microsoft.identity.client.exception.MsalException;

import androidx.annotation.NonNull;

public class Results {
    static class MsalMobileResult {
        private boolean isSuccess;
        private ExceptionDetail exception;
        private ExceptionDetail innerException;
        private Object payload;
        private boolean isUiRequired;

        private MsalMobileResult(@NonNull Exception ex) {
            isSuccess = false;
            exception = new ExceptionDetail(ex);
            if (ex.getCause() != null) {
                innerException = new ExceptionDetail(ex.getCause());
            }
        }

        private MsalMobileResult(@NonNull Object successPayload) {
            isSuccess = true;
            payload = successPayload;
        }

        static String success(@NonNull final Object successPayload) {
            MsalMobileResult result = new MsalMobileResult(successPayload);
            return result.toJson();
        }

        static String error(@NonNull final Exception ex) {
            MsalMobileResult result = new MsalMobileResult(ex);
            return result.toJson();
        }

        static String uiRequiredError(@NonNull final Exception ex) {
            MsalMobileResult result = new MsalMobileResult(ex);
            result.isUiRequired = true;
            return result.toJson();
        }

        private String toJson() {
            Gson gson = new Gson();
            return gson.toJson(this);
        }
    }

    static class ExceptionDetail {
        private String message;
        private String errorCode;

        ExceptionDetail(@NonNull final Throwable throwable) {
            init(throwable);
        }

        ExceptionDetail(@NonNull final Exception exception) {
            init(exception);
        }

        private void init(@NonNull final Throwable throwable) {
            message = throwable.getMessage();

            if (throwable instanceof MsalException) {
                final MsalException ex = (MsalException) throwable;
                errorCode = ex.getErrorCode();
            } else if (throwable instanceof MsalMobileException) {
                final MsalMobileException ex = (MsalMobileException) throwable;
                errorCode = ex.getErrorCode();
            }
        }
    }
}

File: ios/Classes/Results.swift
//
//  Results.swift
//  msal_mobile
//
//  Created by Greg Burke on 12/28/19.
//

import Foundation
import MSAL

class MsalMobileResult : Codable {
    private var isSuccess : Bool
    private var exception : ExceptionDetail?
    private var innerException : ExceptionDetail?
    private var payload : MsalMobileResultPayload?
    public var isUiRequired : Bool
    
    init(exceptionDetail : ExceptionDetail) {
        isSuccess = false
        exception = exceptionDetail
        innerException = nil
        payload = nil
        isUiRequired = false
    }
    
    init(error : Error) {
        isSuccess = false
        exception = ExceptionDetail(error: error)
        innerException = nil // TODO
        payload = nil
        isUiRequired = false
    }
    
    init(successPayload : MsalMobileResultPayload) {
        isSuccess = true
        exception = nil
        innerException = nil
        payload = successPayload
        isUiRequired = false
    }
    
    static func success(successPayload : MsalMobileResultPayload) throws -> String {
        let result = MsalMobileResult(successPayload: successPayload)
        return try result.toJson()
    }
    
    static func error(exception : Error) throws -> String {
        let result = MsalMobileResult(error: exception)
        return try result.toJson()
    }
    
    static func uiRequiredError(exception : Error) throws -> String {
        let result = MsalMobileResult(error: exception)
        result.isUiRequired = true
        return try result.toJson()
    }
    
    func toJson() throws -> String {
        let jsonEncoder = JSONEncoder()
        let jsonData = try jsonEncoder.encode(self)
        let json = String(data: jsonData, encoding: .utf8)
        return json ?? ""
    }
}

class ExceptionDetail : Codable {
    private var message : String
    private var errorCode : String
    
    init(errorCode: String, message: String) {
        self.message = message
        self.errorCode = errorCode
    }
    
    init(error : Error) {
        message = error.localizedDescription
        errorCode = "unknown"
    }
}
